include/matx/transforms/matmul/matvec_cusparse.h

```diff
@@ -0,0 +1,323 @@
+    static_assert(is_tensor_view_v<TensorTypeC>);
+    cudaDataType dtb = MatXTypeToCudaType<TB>();
+    cudaDataType dtc = MatXTypeToCudaType<TC>();
+    ret = cusparseCreateDnVec(&vecB_, params_.n, params_.ptrB, dtb);
+    MATX_ASSERT(ret == CUSPARSE_STATUS_SUCCESS, matxMatMulError);
+    ret = cusparseCreateDnVec(&vecC_, params_.m, params_.ptrC, dtc);
+    MATX_ASSERT(ret == CUSPARSE_STATUS_SUCCESS, matxMatMulError);
+
+    // Allocate a workspace for SpMV.
+    const cusparseSpMVAlg_t algo = CUSPARSE_SPMV_ALG_DEFAULT;
+    const cudaDataType comptp = dtc; // TODO: support separate comp type?!
+    ret =
+        cusparseSpMV_bufferSize(handle_, params_.opA, &salpha_, matA_, vecB_,
+                                &sbeta_, vecC_, comptp, algo, &workspaceSize_);
+    MATX_ASSERT(ret == CUSPARSE_STATUS_SUCCESS, matxMatMulError);
+    if (workspaceSize_) {
+      matxAlloc((void **)&workspace_, workspaceSize_, MATX_DEVICE_MEMORY);
+    }
+  }
+
+  ~MatVecCUSPARSEHandle_t() {
+    if (workspaceSize_) {
+      matxFree(workspace_);
+    }
+    cusparseDestroy(handle_);
+  }
+
+  static detail::MatVecCUSPARSEParams_t
+  GetSpMVParams(TensorTypeC &c, const TensorTypeA &a, const TensorTypeB &b,
+                cudaStream_t stream, float alpha, float beta) {
+    detail::MatVecCUSPARSEParams_t params;
+    params.dtype = TypeToInt<typename TensorTypeA::val_type>();
+    params.ptype = TypeToInt<typename TensorTypeA::pos_type>();
+    params.ctype = TypeToInt<typename TensorTypeA::crd_type>();
+    params.stream = stream;
+    params.alpha = alpha;
+    params.beta = beta;
+    // TODO: simple no-batch, row-wise, no-transpose for now
+    params.nse = a.Nse();
+    params.m = a.Size(TensorTypeA::Rank() - 2);
+    params.n = a.Size(TensorTypeA::Rank() - 1);
+    params.opA = CUSPARSE_OPERATION_NON_TRANSPOSE;
+    // Matrix handles in cuSPARSE are data specific. Therefore, the pointers
+    // to the underlying buffers are part of the GEMM parameters.
+    params.ptrA0 = a.Data();
+    params.ptrA1 = a.POSData(0);
+    params.ptrA2 = a.POSData(1);
+    params.ptrA3 = a.CRDData(0);
+    params.ptrA4 = a.CRDData(1);
+    params.ptrB = b.Data();
+    params.ptrC = c.Data();
+    return params;
+  }
+
+  __MATX_INLINE__ void Exec([[maybe_unused]] TensorTypeC &c,
+                            [[maybe_unused]] const TensorTypeA &a,
+                            [[maybe_unused]] const TensorTypeB &b) {
+    MATX_NVTX_START("", matx::MATX_NVTX_LOG_INTERNAL);
+    const cusparseSpMVAlg_t algo = CUSPARSE_SPMV_ALG_DEFAULT;
+    const cudaDataType comptp = MatXTypeToCudaType<TC>(); // TODO: see above
+    [[maybe_unused]] cusparseStatus_t ret =
+        cusparseSpMV(handle_, params_.opA, &salpha_, matA_, vecB_, &sbeta_,
+                     vecC_, comptp, algo, workspace_);
+    MATX_ASSERT(ret == CUSPARSE_STATUS_SUCCESS, matxMatMulError);
+  }
+
+private:
+  cusparseHandle_t handle_ = nullptr; // TODO: share handle globally?
```

Comment:
We typically never want to share a handle globally since it could cause unknown issues like stalls if there are locks behind using it.

Reply:
Too bad. It would save some overhead for each operation setup (not only SpMV, but all the others too)

Reply:
To be clear, this *may* work, but you'd have to ask the cusparse authors.